### PR TITLE
Optimize test build.

### PR DESCRIPTION
**Pull request by banasraf**
Why we need this PR?

   We want to avoid downloading packages during CI build.

What happend in this PR?

   Script for listing pip package configurations was extended to provide a list of all versions of all packages needed to be predownloaded. All pip installs now use /pip-packages directory to find predownloaded packages. 

Signed-off-by: Rafal <Banas.Rafal97@gmail.com>

### CODE REVIEW

File: qa/setup_packages.py [hunk 1]
```diff
@@ -105,6 +106,22 @@ def cal_num_of_configs(use, cuda):
...
 
+def get_all_string(use, cuda):
+    ret = []
+    for key in packages.keys():
+        if key not in use:
+            continue
+        for val in get_package(packages, key, cuda):
+            if val is None:
+                ret.append(key)
+            elif 'http' in val:
```
Comment (mzient) [inline qa/setup_packages.py]: What is that supposed to do? Won't it be triggered for any val that happens to contain `http` as a substring?

Comment (banasraf) [reply qa/setup_packages.py]: It is copied from `get_install_string`. It differs versions given as urls from those expressed with numbers. It actually needs to handle only those listed in `packages` in this script and it works for them.

Comment (JanuszL) [reply qa/setup_packages.py]:
```suggestion
            elif val.startswith('http'):
```
`get_install_string` could be adjusted as well

File: docker/Dockerfile.customopbuilder.clean [hunk 2]
```diff
@@ -56,4 +56,11 @@ RUN rm -f /usr/bin/python && \
     pip --version
 
 COPY --from=cuda /usr/local/cuda /usr/local/cuda
-ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs/:${LD_LIBRARY_PATH}
\ No newline at end of file
+ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs/:${LD_LIBRARY_PATH}
+
+RUN export USE_CUDA_VERSION=$(cat /usr/local/cuda/version.txt | head -1 | sed 's/.*Version \([0-9]\+\)\.\([0-9]\+\).*/\1/') &&\
+    if [ "${USE_CUDA_VERSION}" = "9" ]; then export SUPPORTED_TF_VERSIONS="1.7.0 1.11.0 1.12.0"; fi && \
```
Comment (klecki) [inline docker/Dockerfile.customopbuilder.clean]: I think, we should have single source of truth for supported TensorFlow versions. Now it is listed here and in `qa/setup_packages.py`.

Comment (JanuszL) [reply docker/Dockerfile.customopbuilder.clean]: They doesn't have to match. In test we may have less (just a few old, and the most recent), here we need all that differs in ABI.

Comment (klecki) [reply docker/Dockerfile.customopbuilder.clean]: Currently they are just copy-pasted.

Comment (klecki) [reply docker/Dockerfile.customopbuilder.clean]: But if they are really independent, that's ok.

Comment (JanuszL) [reply docker/Dockerfile.customopbuilder.clean]: Now they are:
- https://github.com/NVIDIA/DALI/blob/master/qa/setup_packages.py
- this docker file
- https://github.com/NVIDIA/DALI/blob/master/dali_tf_plugin/build_in_custom_op_docker.sh
Maybe one common file inside dali_tf_plugin that will be read/included in all other places could work.
What do you think?

File: docker/Dockerfile.customopbuilder.clean [hunk 3]
```diff
@@ -56,4 +56,12 @@ RUN rm -f /usr/bin/python && \
...
 
 COPY --from=cuda /usr/local/cuda /usr/local/cuda
-ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs/:${LD_LIBRARY_PATH}
\ No newline at end of file
+ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs/:${LD_LIBRARY_PATH}
+
+WORKDIR /opt/dali
+COPY qa/setup_packages.py qa/setup_packages.py
+RUN export USE_CUDA_VERSION=$(cat /usr/local/cuda/version.txt | head -1 | sed 's/.*Version \([0-9]\+\)\.\([0-9]\+\).*/\1\2/') &&\
+    export last_config_index=$(python qa/setup_packages.py -n -u tensorflow-gpu --cuda ${USE_CUDA_VERSION}) &&\
```
Comment (jantonguirao) [inline docker/Dockerfile.customopbuilder.clean]: nitpick (several occurrences) 
```suggestion
    export last_config_index=$(python qa/setup_packages.py -n -u tensorflow-gpu --cuda ${USE_CUDA_VERSION}) && \
```

File: qa/setup_packages.py [hunk 4]
```diff
@@ -105,6 +106,22 @@ def cal_num_of_configs(use, cuda):
         ret *= len(get_package(packages, key, cuda))
     return ret
 
+def get_all_string(use, cuda):
```
Comment (jantonguirao) [inline qa/setup_packages.py]:
```suggestion
def get_all_strings(use, cuda):
```

File: docker/Dockerfile.customopbuilder.clean (hunk 2 above)
Comment (JanuszL) [reply docker/Dockerfile.customopbuilder.clean]: Done

File: qa/setup_packages.py (hunk 4 above)
Comment (JanuszL) [reply qa/setup_packages.py]: Done

File: docker/Dockerfile.customopbuilder.clean (hunk 3 above)
Comment (JanuszL) [reply docker/Dockerfile.customopbuilder.clean]: Done

File: docker/Dockerfile.customopbuilder.clean [hunk 5]
```diff
@@ -56,4 +56,12 @@ RUN rm -f /usr/bin/python && \
...
 
 COPY --from=cuda /usr/local/cuda /usr/local/cuda
-ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs/:${LD_LIBRARY_PATH}
\ No newline at end of file
+ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs/:${LD_LIBRARY_PATH}
+
+WORKDIR /opt/dali
+COPY qa/setup_packages.py qa/setup_packages.py
+RUN export USE_CUDA_VERSION=$(cat /usr/local/cuda/version.txt | head -1 | sed 's/.*Version \([0-9]\+\)\.\([0-9]\+\).*/\1\2/') && \
+    export last_config_index=$(python qa/setup_packages.py -n -u tensorflow-gpu --cuda ${USE_CUDA_VERSION}) && \
```
Comment (szalpal) [inline docker/Dockerfile.customopbuilder.clean]: Could you put together one line of comment, what this layer is supposed to do? It's quite enigmatic with all the `sed` calls and for loops ;)

Comment (JanuszL) [reply docker/Dockerfile.customopbuilder.clean]: Done

**Pull request merged**